chronos/timer.nim

```diff
@@ -8,45 +8,397 @@
+  ## Moment += Duration
+  a.value += b.value
+
+func `+=`*(a: var Duration, b: Duration) {.inline.} =
+  ## Duration += Duration
+  a.value += b.value
+
+func `-`*(a, b: Moment): Duration {.inline.} =
+  ## Moment - Moment = Duration
+  ##
+  ## Note: Duration can't be negative.
+  result.value = if a.value >= b.value: a.value - b.value else: 0'i64
+
+func `-`*(a: Moment, b: Duration): Moment {.inline.} =
+  ## Moment - Duration = Moment
+  ##
+  ## Note: Moment can be negative
+  result.value = a.value - b.value
+
+func `-`*(a: Duration, b: Duration): Duration {.inline.} =
+  ## Duration - Duration = Duration
+  ##
+  ## Note: Duration can't be negative.
+  result.value = if a.value >= b.value: a.value - b.value else: 0'i64
+
+func `-=`*(a: var Duration, b: Duration): Duration {.inline.} =
+  ## Duration -= Duration
+  a.value = if a.value >= b.value: a.value - b.value else: 0'i64
+
+func `-=`*(a: var Moment, b: Duration): Moment {.inline.} =
+  ## Moment -= Duration
+  a.value -= b.value
+
+func `==`*(a, b: Duration): bool {.inline.} =
+  ## Returns ``true`` if ``a`` equal to ``b``.
+  result = (a.value == b.value)
+
+func `==`*(a, b: Moment): bool {.inline.} =
+  ## Returns ``true`` if ``a`` equal to ``b``.
+  result = (a.value == b.value)
+
+func `<`*(a, b: Duration): bool {.inline.} =
+  ## Returns ``true`` if ``a`` less then ``b``.
+  result = (a.value < b.value)
+
+func `<`*(a, b: Moment): bool {.inline.} =
+  ## Returns ``true`` if ``a`` less then ``b``.
+  result = (a.value < b.value)
+
+func `<=`*(a, b: Duration): bool {.inline.} =
+  ## Returns ``true`` if ``a`` less or equal ``b``.
+  result = (a.value <= b.value)
+
+func `<=`*(a, b: Moment): bool {.inline.} =
+  ## Returns ``true`` if ``a`` less or equal ``b``.
+  result = (a.value <= b.value)
+
+func `>`*(a, b: Duration): bool {.inline.} =
+  ## Returns ``true`` if ``a`` bigger then ``b``.
+  result = (a.value > b.value)
+
+func `>`*(a, b: Moment): bool {.inline.} =
+  ## Returns ``true`` if ``a`` bigger then ``b``.
+  result = (a.value > b.value)
+
+func `>=`*(a, b: Duration): bool {.inline.} =
+  ## Returns ``true`` if ``a`` bigger or equal ``b``.
+  result = (a.value >= b.value)
+
+func `>=`*(a, b: Moment): bool {.inline.} =
+  ## Returns ``true`` if ``a`` bigger or equal ``b``.
+  result = (a.value >= b.value)
+
+func `*`*(a, b: Duration): Duration {.inline.} =
+  result.value = a.value * b.value
+
+func `div`*(a, b: Duration): Duration {.inline.} =
+  result.value = a.value div b.value
+
+const
+  Nanosecond* = Duration(value: 1'i64)
+  Microsecond* = Duration(value: 1_000'i64) * Nanosecond
+  Millisecond* = Duration(value: 1_000'i64) * Microsecond
+  Second* = Duration(value: 1_000'i64) * Millisecond
+  Minute* = Duration(value: 60'i64) * Second
+  Hour* = Duration(value: 60'i64) * Minute
+  Day* = Duration(value: 24'i64) * Hour
+  Week* = Duration(value: 7'i64) * Day
+
+  ZeroDuration* = Duration(value: 0'i64)
+  InfiniteDuration* = Duration(value: -1'i64)
```

Comment:
I think it might be better to have this as `high(int64)` - that will make it sort naturally and we can saturate operations that would otherwise overflow to this value (just like we saturate the lower bound to 0)

Reply:
high(int64) is still valid Duration, while `-1` is marked as Infinite, its normally to define it out of range.

Reply:
yes, so it's a question of which is more useful - simple sorting or that last value (out of 2^63), when talking about nanoseconds. I think sorting, in this particular case, because we're using durations for timeouts often, so we want to sort many lists and it seems useful to have the inifinate ones at the end... what are good use cases for that last value?

ie consider the windows DWORD infinate - they use 0xffffffff which is in-range technically, and sorts correctly.